### PR TITLE
Add a new test for updateRenderState()

### PR DESCRIPTION
**Pull request by svillar**
This PR adds a new test for the updateRenderState() algorithm described https://immersive-web.github.io/webxr/#dom-xrsession-updaterenderstate.

### CODE REVIEW

File: webxr/render_state_update.https.html
Comment (Manishearth) [inline webxr/render_state_update.https.html:32]: can we just assert_throws this?

File: webxr/render_state_update.https.html
Comment (Manishearth) [inline webxr/render_state_update.https.html:28]: A correctly implemented UA will not let you create a second immersive session in the first place, this can only work for two inline sessions

File: webxr/render_state_update.https.html
Comment (svillar) [reply webxr/render_state_update.https.html:32]: ok

File: webxr/render_state_update.https.html
Comment (svillar) [reply webxr/render_state_update.https.html:28]: good point, will change it.

**Comment by svillar** [comment]
@Manishearth mind taking another look? Thanks!

File: webxr/render_state_update.https.html
Comment (Manishearth) [reply webxr/render_state_update.https.html:28]: Right, but you need to end the first session, no? I don't see that happening here.

File: webxr/render_state_update.https.html
Comment (Manishearth) [inline webxr/render_state_update.https.html:16]: assert_throws please

File: webxr/render_state_update.https.html
Comment (Manishearth) [inline webxr/render_state_update.https.html:42]: assert_throws please

File: webxr/render_state_update.https.html
Comment (svillar) [reply webxr/render_state_update.https.html:28]: Well I might be missunderstanding the specs, but in the requestSession() algorithm in the step 5.4.8 just before resolving the promise it says:

>  If immersive is true
> Set the active immersive session to session, and set pending immersive session to false.

So there is no need to end the session to be allowed to request another one.

Comment (Manishearth) [reply webxr/render_state_update.https.html:28]: Look further up, in step 4, "if immersive is true", substep 2:

> If pending immersive session is true or active immersive session is not null, reject promise with an "InvalidStateError" DOMException and return promise.

Comment (svillar) [reply webxr/render_state_update.https.html:28]: > Look further up, in step 4, "if immersive is true", substep 2:
> 
> > If pending immersive session is true or active immersive session is not null, reject promise with an "InvalidStateError" DOMException and return promise.

I know. The thing is that I request the second session when the first session is already resolved. So, just before resolving, the first session sets "pending immersive session" to false, so when the second request arrives the step 4 substep 2 you mention does not apply.

Comment (Manishearth) [reply webxr/render_state_update.https.html:28]: Right, when the session is resolved the pending immersive session is set to false, but the active immersive session is set to the new session, and that's also checked in 4.2.

The spec should not allow simultaneous immersive sessions to exist. That's why there's a single active immersive session field. If the spec is somehow allowing this, that's a bug (but I still fail to see how it does), but either way, there can only ever be a single immersive session.

Comment (svillar) [reply webxr/render_state_update.https.html:28]: You're right. Actually we do check that in WebKit code, so there must be a bug somewhere else. Thanks for the review, I'll upload a new version after figuring out why it isn't failing.

Comment (svillar) [reply webxr/render_state_update.https.html]: The code was OK. The problem was the test. The exception in both cases (requesting to immersive sessions or creating a base layer with a different sessions) is `InvalidStateError`, so the test was actually capturing the former instead of the latter.

**Review by Manishearth — APPROVED** [review]
cc @alcooper91

File: webxr/render_state_update.https.html
Comment (alcooper91) [inline webxr/render_state_update.https.html:34]: This test already exists via: webxr/render_state_vertical_fov_inline.https.html

File: webxr/render_state_update.https.html
Comment (alcooper91) [inline webxr/render_state_update.https.html:67]: NIT: Copy paste error, please update

File: webxr/render_state_update.https.html
Comment (alcooper91) [reply webxr/render_state_update.https.html:34]: Ah, and also throwing isn't technically compliant: https://immersive-web.github.io/webxr/#apply-the-pending-render-state

File: webxr/render_state_update.https.html
Comment (svillar) [reply webxr/render_state_update.https.html:67]: Ups, I'll fix that.

File: webxr/render_state_update.https.html
Comment (svillar) [reply webxr/render_state_update.https.html:34]: I think they're testing two different things. I'm trying to test the updateRenderState() API (which indeed throws) while the test you mention is actually testing the apply render state if I am not wrong

Comment (alcooper91) [reply webxr/render_state_update.https.html:34]: I disagree, step 11 from that algorithm https://immersive-web.github.io/webxr/#dom-xrsession-updaterenderstate
Doesn't mention any validation to inlineVerticalFieldOfView except for throwing if it's set and being applied to an immersive session.
Applying the pending render state is where any validation actually happens, and that happens by clamping.

Comment (alcooper91) [reply webxr/render_state_update.https.html:34]: Actually... I just looked closer, and this *is* setting to an immersive session (Not sure why I assumed it was setting on an inline session...).

That is also already tested via: webxr/render_state_vertical_fov_immersive.https.html

**Comment by alcooper91** [comment]
General comment, please wrap all asserts in t.step(() => { // foo };
I believe that failures don't get reported to the testharness cleanly otherwise.

That's probably something we should look into in webxr_util.js at a future date (I think it has to do with how we resolve the function, but it may be an unavoidable side-effect of how the tests assert within a promise/callback that *isn't* directly part of the chain...)

File: webxr/render_state_update.https.html
Comment (svillar) [reply webxr/render_state_update.https.html:34]: > I disagree, step 11 from that algorithm https://immersive-web.github.io/webxr/#dom-xrsession-updaterenderstate
> Doesn't mention any validation to inlineVerticalFieldOfView except for throwing if it's set and being applied to an immersive session.
> Applying the pending render state is where any validation actually happens, and that happens by clamping.

Then I might be missunderstanding the specs. What I am trying to test is step 4 of `updateRenderState()`, that's before the step 11 you mention.

**Comment by svillar** [comment]
> General comment, please wrap all asserts in t.step(() => { // foo };
> I believe that failures don't get reported to the testharness cleanly otherwise.
>
Sure, I'll do that. Thanks for the suggestion.

File: webxr/render_state_update.https.html
Comment (svillar) [reply webxr/render_state_update.https.html:34]: Just to be crystal clear, I don't want to test the validation at all (as you said there are 2 other tests for that), I just want to test the first 5 steps of the update algorithm.

Comment (alcooper91) [reply webxr/render_state_update.https.html:34]: Yeah; sorry, my initial understanding was wrong and I thought it was the step 11; however step 4 *is* still tested (though it doesn't use assert_throws and ~implements that itself) in this test: https://github.com/web-platform-tests/wpt/blob/master/webxr/render_state_vertical_fov_immersive.https.html
(Similar name to the one I initially linked, but replicates this test).

Comment (svillar) [reply webxr/render_state_update.https.html:34]: OK, I understand that you prefer to remove it then. I'll understand silence as a confirmation :)

**Comment by svillar** [comment]
Thanks for the patience. I was on holidays so I couldn't update this before. PTAL.

File: webxr/render_state_update.https.html
Comment (alcooper91) [inline webxr/render_state_update.https.html:61]: NitPick: Only the asserts really need to be in the t.step, though I don't think this hurts anything.

**Review by alcooper91 — APPROVED** [review]
Thanks!

File: webxr/render_state_update.https.html
Comment (svillar) [reply webxr/render_state_update.https.html:61]: I just basically did it because I saw other tests with whole code blocks inside it.

**Comment by svillar** [comment]
TC reports that `wpt-decision-task` is pending but when clicking on "Details" it says it was completed. What's going on?

**Comment by svillar** [comment]
@Manishearth @alcooper91 do we really need @klausw's and @toji's approvals?

**Comment by Manishearth** [comment]
Alex's review should be fine

**Comment by svillar** [comment]
> Alex's review should be fine

Any idea why some checks are not apparently completed?

**Review by Manishearth — APPROVED** [review]

**Comment by Manishearth** [comment]
Alex might not have merge access to this repo? But I do, and even with my review it's not working

**Comment by Manishearth** [comment]
Retriggered the build

**Comment by svillar** [comment]
> Retriggered the build

Seems like it didn't work either

**Comment by Manishearth** [comment]
Cc @stephenmcgruer ?

**Comment by stephenmcgruer** [comment]
Huh. This looks like it could be a Taskcluster bug - the actual task succeeded (https://community-tc.services.mozilla.com/tasks/groups/MKIoi4r8QxyD-pea0sz9kw are all green) but Taskcluster isn't reporting the status to GitHub... thanks for the tag - I'll admin merge and keep an eye out for more reports of this nature.

**Comment by svillar** [comment]
> Huh. This looks like it could be a Taskcluster bug - the actual task succeeded (https://community-tc.services.mozilla.com/tasks/groups/MKIoi4r8QxyD-pea0sz9kw are all green) but Taskcluster isn't reporting the status to GitHub... thanks for the tag - I'll admin merge and keep an eye out for more reports of this nature.

Thanks @stephenmcgruer !

**Pull request merged**
